File: test-suite/golang/security/redos_regex_clean.go
package security

import (
	"net/http"
	"regexp"
)

func allowedRegexpPattern(kind string) string {
	switch kind {
	case "email":
		return `^[^@]+@example\.com$`
	case "status":
		return `^(active|archived)$`
	default:
		return `^$`
	}
}

func escapedQueryRegex(r *http.Request) (*regexp.Regexp, error) {
	raw := r.URL.Query().Get("pattern")
	pattern := regexp.QuoteMeta(raw)
	return regexp.Compile(pattern)
}

func allowlistedRegex(r *http.Request) *regexp.Regexp {
	raw := r.URL.Query().Get("kind")
	pattern := allowedRegexpPattern(raw)
	return regexp.MustCompile(pattern)
}

func fixedPattern(value string) bool {
	matched, _ := regexp.MatchString(`^(active|archived)$`, value)
	return matched
}

File: test-suite/js/security/redos-regex-clean.ts
type RequestLike = {
  query: Record<string, string | undefined>;
  nextUrl: { searchParams: URLSearchParams };
};

const FIXED_STATUS_PATTERN = /^(active|archived)$/;

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function allowedRegexPattern(value: string): string {
  switch (value) {
    case "email":
      return "^[^@]+@example\\.com$";
    case "status":
      return "^(active|archived)$";
    default:
      throw new Error("unsupported pattern");
  }
}

export function escapedQueryPattern(req: RequestLike): RegExp {
  const raw = req.query.pattern ?? "";
  return new RegExp(escapeRegExp(raw), "i");
}

export function allowlistedPattern(request: RequestLike): RegExp {
  const raw = request.nextUrl.searchParams.get("kind") ?? "status";
  const pattern = allowedRegexPattern(raw);
  return new RegExp(pattern);
}

export function fixedPattern(): RegExp {
  return FIXED_STATUS_PATTERN;
}
